PolyEngine/Core/Src/String.cpp

```diff
@@ -0,0 +1,207 @@
+
+using namespace Poly;
+
+size_t Poly::StrLen(const char* str) {
+	size_t len = 0;
+	while (str[len] != 0)
+		++len;
+	return len;
+}
+
+
+String::String(const char* data) {
+	size_t length = StrLen(data);
+	Data.Resize(length + 1);
+	std::memcpy(Data.GetData(), data, sizeof(char) * length);
+	Data[length] = 0;
+}
+
+String::String(const String& rhs) {
+	*this = rhs;
+}
+
+String::String(String&& rhs) {
+	*this = String(std::move(rhs));
+}
+
+
+String* String::From(int var) {
+	return new String(std::to_string(var).c_str());
+}
+
+String* String::From(float var) {
+	return new String(std::to_string(var).c_str());
+}
+
+String* String::From(double var) {
+	return new String(std::to_string(var).c_str());
+}
+
+String* String::From(char var) {
+	return From(std::string(1, var));
+}
+
+String* String::From(const char* var) {
+	return new String(var);
+}
+
+String* String::From(const std::string& var) {
+	return new String(var.c_str());
+}
+
+
+bool String::Contains(const String& var) const {
+	//TODO Dynarray.Contains(Dynarray<T> rhs)
+	return false;
+}
+
+String* String::ToLower() const {
+	Dynarray<char> d = Dynarray<char>(this->Data.GetSize());
+	for (int i = 0; i < this->Data.GetSize(); i++) {
+		char c = this->Data[i];
+		if (c > 64 && c < 91) {
+			c += 32;
+		}
+		d.PushBack(c);
+	}
+	return new String(d.GetData());
+}
+
+String* String::ToUpper() const {
+	Dynarray<char> d = Dynarray<char>(this->Data.GetSize());
+	for (int i = 0; i < this->Data.GetSize(); i++) {
+		char c = this->Data[i];
+		if (c > 96 && c < 123) {
+			c -= 32;
+		}
+		d.PushBack(c);
+	}
+	return new String(d.GetData());
+}
+
+bool String::IsEmpty() const {
+	return (bool)GetLength();
+}
+
+String* String::Replace(char a, char b) const {
+	Dynarray<char> d = Dynarray<char>(this->Data.GetSize());
+	for (int i = 0; i < this->Data.GetSize(); i++) {
+		char c = this->Data[i];
+		if (c == a) {
+			c = b;
+		}
+		d.PushBack(c);
+	}
+	return new String(d.GetData());
+}
+
+String* String::Replace(const String& a, const String& b) const {
+	//TODO Dynarray.Contains(Dynarray<T> rhs)
+	return nullptr;
+}
+
+String* String::Split(char delimiter) const {
+	//TODO Dynarray.FindAllIdx(T rhs)
+	return nullptr;
+}
+
+String* String::Join(const String& vars, size_t size, const String& separator) {
+	//TODO stringbuilder
+	return nullptr;
+}
+
+bool String::StartsWith(char var) const {
+	return (Data[0] == var);
+}
+
+bool String::EndsWith(char var) const {
+	return (Data[Data.GetSize()-2] == var);
+}
+
+String* String::Substring(int end) const {
+	return Substring(0, end);
+}
+
+String* String::Substring(int start, int end) const {
+	//TODO dynarray constructor?
+	Dynarray<char> d = Dynarray<char>(end - start + 1);
+	for (int i = start; i < end; i++) {
+		d.PushBack(this->Data[i]);
+	}
+	d.PushBack(0); //why is this needed? constructor should do that
+	return new String(d.GetData());
+}
+
+
+String* String::GetTrimmed() const {
+	//TODO substrings
+	Dynarray<char> d = this->Data;
```

Comment:
Instead of dynarray manipulation find startIdx and endIdx of the trimmed string, and call SubStr.